src/library/lazy.lisp

```diff
@@ -0,0 +1,57 @@
+(in-package #:coalton-library)
+
+(cl:defmacro lazy (body)
```

Comment:
This doesn't need to be a macro... A Coalton function would probably be better suited so it can be typechecked

Reply:
If it's not a macro, the user has to wrap it in a function themselves to avoid evaluating the expression, which feels a bit unergonomic.

Reply:
I would prefer not to have macros which automatically wrap code in lambdas because it obscures control flow. I also think we should keep macros to a minimum because they make Coalton's errors more difficult to understand

Reply:
> I would prefer not to have macros which automatically wrap code in lambdas because it obscures control flow. I also think we should keep macros to a minimum because they make Coalton's errors more difficult to understand

I agree on both points, but I think `lazy` is a good use-case. Supplying an anonymous function is boilerplate and expose implementation details. `lazy` as a language-level construct should already be familiar to people coming from scheme (`delay`) and F# (`lazy`), and probably others (guess `clazy` for CL does the same thing). C# needs to supply a lambda to construct a `Lazy<T>`.

Reply:
> A Coalton function would probably be better suited so it can be typechecked

The typechecker still gives acceptable error messages

```lisp
  ;; Failed to unify types STRING and INTEGER
  ;; in unification of types (LAZY STRING) and (LAZY INTEGER)
  ;; in definition of X

  (declare x (Lazy String))
  (define x (lazy 1))
```